pkg/datapath/linux/routing/migrate_test.go

```diff
@@ -0,0 +1,148 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build privileged_tests
+
+package linuxrouting
+
+import (
+	"fmt"
+	"net"
+	"os/exec"
+	"runtime"
+
+	"github.com/cilium/cilium/pkg/datapath/linux/linux_defaults"
+
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netns"
+	. "gopkg.in/check.v1"
+)
+
+var _ = Suite(&MigrateSuite{})
+
+type MigrateSuite struct {
+	// interfaceDB interface mock
+	OnGetInterfaceNumberByMAC func(mac string) (int, error)
+}
+
+func (m *MigrateSuite) TestMigrateENIRoutingTable(c *C) {
+	// This is required to make Golang play safe with netns's.
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	// First, we need to setup the Linux routing policy database to mimic a
+	// broken setup. Then we will call MigrateENIRoutingTable.
+
+	previousNetNS, err := netns.Get()
+	c.Assert(err, IsNil)
+	defer func() { c.Assert(netns.Set(previousNetNS), IsNil) }()
+
+	newNetNS, err := netns.New()
+	c.Assert(err, IsNil)
+	c.Assert(int(newNetNS), Not(Equals), -1)
+
+	m.assertRoutingTable(c)
+}
+
+func (m *MigrateSuite) assertRoutingTable(c *C) {
+	// (1) Setup the routing table.
+
+	// Pick an arbitrary iface index because that's what we previously used as
+	// a table ID (see RoutingInfo). All the old rules and routes will be set
+	// up with this table ID.
+	//
+	// Pick an arbitrary interface number. The new table ID will be 10
+	// (linux_defaults.RouteTableInterfacesOffset) + this number (= 11).
+	index := 5
+	ifaceNumber := 1
+	newTableID := 11
+
+	// (1) Setting up the routing table for testing upgrading.
+
+	dummyTmpl := &netlink.Dummy{
+		LinkAttrs: netlink.LinkAttrs{
+			Name:  "gotestdummy0",
+			Index: index,
+		},
+	}
+	c.Assert(netlink.LinkAdd(dummyTmpl), IsNil)
+	c.Assert(netlink.LinkSetUp(dummyTmpl), IsNil)
+	c.Assert(netlink.AddrAdd(dummyTmpl, &netlink.Addr{
+		IPNet: &net.IPNet{IP: net.ParseIP("172.2.0.2"), Mask: net.CIDRMask(24, 32)},
+	}), IsNil)
+	c.Assert(netlink.RouteAdd(&netlink.Route{
+		Dst:   &net.IPNet{IP: net.IPv4zero, Mask: net.CIDRMask(0, 32)},
+		Gw:    net.ParseIP("172.2.0.1"),
+		Table: dummyTmpl.Index,
+	}), IsNil)
+
+	// _, cidr, err := net.ParseCIDR("172.2.0.2/32")
+	// c.Assert(err, IsNil)
+	// c.Assert(netlink.RuleAdd(&netlink.Rule{
+	// 	// Src:      &net.IPNet{IP: net.ParseIP("172.2.0.2"), Mask: net.CIDRMask(32, 32)},
+	// 	Src: cidr,
+	// 	// Dst:      &net.IPNet{IP: net.IPv4zero, Mask: net.CIDRMask(0, 32)},
+	// 	Priority: linux_defaults.RulePriorityEgress,
+	// 	Table:    dummyTmpl.Index,
+	// }), IsNil)
+
+	// TODO(christarazi): The above is commented out because it fails with
+	// EINVAL on my machine. Below is a workaround.
+	err := exec.Command("ip", "rule", "add",
+		"from", "172.2.0.2/32",
+		"to", "all",
+		"table", fmt.Sprintf("%d", index),
+		"priority", fmt.Sprintf("%d", linux_defaults.RulePriorityEgress)).Run()
+	c.Assert(err, IsNil)
```

Comment:
Not sure what to do with this

Reply:
I'm genuinely interested in the underlying issue here, but as far as the immediate problem of writing test validation code, I believe `ip` is typically available in all environments where we run the privileged tests so I'm fine with just sticking with the uncommented implementation and we can dig separately on it.

FWIW I also tried this locally and found the same behaviour. There's likely some difference between how the library prepares the kernel netlink request vs. what `ip rule` does, and it may even be a netlink library bug or something. But from a pragmatic perspective, I don't think there's any immediate issue with just shelling out here.

Reply:
^ We could file a bug to investigate further and reference it in the code like `// GH-XXXX: Must shell out here due to netlink request resulting in EINVAL.`

Reply:
Done.